stdlib/buffer.gr

```diff
@@ -0,0 +1,575 @@
+import Memory from "runtime/unsafe/memory"
+import WasmI32 from "runtime/unsafe/wasmi32"
+import Conv from "runtime/unsafe/conv"
+import Exception from "runtime/exception"
+import Int32 from "int32"
+import Bytes from "bytes"
+import String from "string"
+
+record Buffer {
+  mut len: Number,
+  initialSize: Number,
+  mut data: Bytes,
+}
+
+let _SIZE_OFFSET = 4n;
+
+let _VALUE_OFFSET = 8n;
+
+let _8BIT_LEN = 1
+
+let _16BIT_LEN = 2
+
+let _32BIT_LEN = 4
+
+let _64BIT_LEN = 8
+
+/** Converts Number to WasmI32 */
+@disableGC
+let numberToWasmI32 = n => {
+  let m = WasmI32.fromGrain(n)
+  if (WasmI32.eqz(WasmI32.and(m, 1n))) {
+    let i32 = Int32.fromNumber(n)
+    let res = Conv.fromInt32(i32)
+    Memory.decRef(WasmI32.fromGrain(i32))
+    res
+  } else {
+    // fast path
+    WasmI32.shrS(m, 1n)
+  }
+}
+
+/** Gets the size of a Bytes via its ptr */
+let getSize = (ptr) => WasmI32.load(ptr, _SIZE_OFFSET)
+
+let checkIndexIsInBounds = (i, len, buf) => {
+  if (i < 0) {
+    throw Exception.IndexOutOfBounds
+  }
+  if (i >= buf.len || i + len > buf.len) {
+    throw Exception.IndexOutOfBounds
+  }
+}
+
+let grow = (buf) => {
+  let mut n = Bytes.length(buf.data)
+  if (n == 0) n = 4 // Make sure bytes of 0 length grow too
+  buf.data = Bytes.resize(0, n, buf.data)
+}
+
+let autogrow = (len, buf) => {
+  let mut limit = 100 // limiting number of sequential capacity doubles to 100
+  while (buf.len + len > Bytes.length(buf.data)) {
+    if (limit <= 0) {
+      throw Exception.Overflow
+    }
+    grow(buf)
+    limit = limit - 1
+  }
+}
+
+@disableGC
+let getCharAsWasmI32 = (char) => {
+  let c = WasmI32.fromGrain(char)
+  WasmI32.load(c, 4n)
+}
+
+@disableGC
+let getCharByteLength = (byte) => {
+  let (+) = WasmI32.add
+  let (&) = WasmI32.and
+  let (==) = WasmI32.eq
+  if ((byte & 0x80n) == 0x00n) {
+    1n
+  } else if ((byte & 0xF0n) == 0xF0n) {
+    4n
+  } else if ((byte & 0xE0n) == 0xE0n) {
+    3n
+  } else {
+    2n
+  }
+}
+
+@disableGC
+let appendBytes = (srcOff, dstOff, len, src, dst) => {
+  let (+) = WasmI32.add
+  Memory.copy(dst + _VALUE_OFFSET + dstOff, src + _VALUE_OFFSET + srcOff, len)
+}
+
+/**
+ * Gets a signed 8-bit integer starting at the given byte index.
+ * @param i: Number - The byte index
+ * @param b: Buffer - The buffer
+ * @returns Int32
+ */
```

Comment:
Need to update these. The syntax for docblocks is currently documented at https://grain-lang.org/blog/2021/08/27/tooling-preview-graindoc/

Here's the gist. I'd also prefer external facing arguments not be 1-letter. We may need to rework my verbiage.
```suggestion
/**
 * Gets a signed 8-bit integer starting at the given byte index.
 *
 * @param index: The byte index to access
 * @param buffer: The buffer to access
 * @returns The 8-bit integer that starts at the given index
 *
 * @since v0.4.0
 */
```